EIPS/eip-3030.md

```diff
@@ -0,0 +1,397 @@
+Content | `{"status": "OK"}`
+
+---
+
+### `GET /publicKeys`
+
+_**Responses**_
+
+Success | <br>
+--- | ---
+Code | `200`
+Content | `{"public_keys": "[public_key_hex_string_without_0x]"}`
+
+---
+
+### `POST /sign/:public_key`
+
+URL Parameter | <br>
+--- | ---
+`:public_key` | `public_key_hex_string_without_0x`
+
+_**Request**_
+
+JSON Body | <br> | <br>
+--- | --- | ---
+`signingRoot` | **Required** | A string representation of the hexadecimal value to be signed
+<br> | Optional | Any other field will be ignored by the signer
+
+_**Responses**_
+
+Success | <br>
+--- | ---
+Code |  `200`
+Content | `{"signature": "<signature_hex_string>"}`
+
+_or_
+
+Error | <br>
+--- | ---
+Code |  `400`
+Content | `{"error": "<Bad Request Error Message>"}`
+
+_or_
+
+Error | <br>
+--- | ---
+Code |  `404`
+Content | `{"error": "Key not found: <public_key_hex_string_without_0x>"}`
+
+---
+
+## Rationale
+
+### URL parameter `:public-key` without `0x`
+
+By relaxing the constraint of requiring hexadecimal input, clients can opt-in to using custom ids for their private keys, and perform the adequate retrieving and matching with the secret key within the signer. That is, if the remote signer implementer allows for this feature.
+
+### UNIX philosophy: Simple API
+
+This API specification contains only three methods: one for **status**, one for **listing the available keys**, and one to **produce a signature**. There are no methods for authentication, pre-image validation, key management, nor transport encryption.
+
+The following subsections discuss aspects to be considered by the client implementers relative to these subjects.
+
+#### Implementation of additional features
+
+From an API pipeline view, we have two nodes: The validator client (1) that makes requests to the remote signer (2). A more sophisticated chain can be built by introducing elements between these two nodes. Either by setting up reverse proxy services, or by adding plugins to the remote signer implementation.
+
+#### Authentication
+
+Can be accomplished through the use of an HTTP Request Header. There are several ways to negotiate and issue a valid token to authenticate the communication between the validator client and the remote signer, each of them with potential drawbacks (e.g replay attacks, challenges in distributing the token to the validator client, etc.). In general, any method of authentication must be combined with transport encryption to be effective.
+
+The operator can also implement network Access Control Lists (ACLs) between the validator client's network and the remote signer's network, reducing the attack surface by requiring a potential attacker to be positioned in the same network as the validator client.
+
+#### Pre-image validation
+
+A key feature for a remote signer, pre-image validation, implies that not only the `signingRoot`, but all the required elements needed to perform complete validation of the message, are sent through the wire to obtain a signature.
```

Comment:
I'd argue that pre-image validation is such a core feature that the post-image shouldn't even be sent to the signer. The signer should always be forced to calculate the hash-to-sign locally, given the input + info about signing request (maybe using mimetypes to signal what type of signing is requested)

Reply:
With this spec, as is, you cement the behaviour of a standard-compliant signer to sign blind checks. Which seems like a very bad idea, IMO

Reply:
We can do the following then:

- Add as **required** fields `domain`, the `type` of message to be produced (randao, block, or attestation), and the specific `data` per type of message.
- It is responsibility of the Remote Signer Client to perform the adequate checks on the received payload in order to admit it and process the signature.

Reply:
You can also _remove_ the `signingRoot` parameter. Btw, can you point me towards info about what the `signingRoot` consists of? Is it the ssz-hash of a full block? or a header? How large is the preimage?

Reply:
The `signingRoot` is the string representation of a SHA256 Hash. The result of the Ethereum 2.0 function [`compute_signing_root`](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/beacon-chain.md#compute_signing_root).

Reply:
I see. So, in practice, what's the size of the `ssz_object` ? Why not send it in it's entirety? 
And depending on the size of that -- if it's extremely large, perhaps only portions + proofs/hashes for missing parts?

Reply:
Fyi this is Prysm's definition. We have a field for the beacon object for additional verifications such as slashing prevention.

https://github.com/prysmaticlabs/prysm/blob/master/proto/validator/accounts/v2/keymanager.proto#L53

It's not a mandatory field, the remote can ignore it

Reply:
there s one scenario in my head where having the signingRoot together with all the needed fields to recompute it might come in handy, tamper detection. if the sent root and the calculated root mismatch then you have a big problem

Reply:
In that case, someone is tampering with the traffic. So if they can modify one (e.g. the signingRoot, to sign something else), why wouldn't they tamper with both? 
On the one hand, there's a vague argument "it might be handy". So the arguments _for_ including it are: 

- It may be handy, against a sloppy attack where the attacker modifies one but not both fields

The arguments _against_ including it are: 

- If it's included, a lazy implementation or insecurely configured signer can choose to blindly sign anything. 
- If we do not include it, we _force_ to sealer to _at least_ verify that it's a block being sealed, and not something else. 
- If we do not include it, since the selaer needs to parse the ssz-object, it becomes a very small step to also do at least some basic anti-slashing validations, such as "do not seal same height twice".

Reply:
well, you are right of course :D thanks , 
pushing for everything to get signer to not sign anything blindly should be the the goal